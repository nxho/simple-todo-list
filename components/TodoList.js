import React from 'react';
import { StyleSheet, Text, View } from 'react-native';

import DateHeader from './DateHeader';
import ItemList from './ItemList';
import AddItemButton from './AddItemButton';

export default class TodoList extends React.Component {
	constructor(props) {
		super(props);

		this.state = {
			items: [],
		}
	}

	handleChange = (text, index) => {
		this.state.items[index].text = text;
		this.setState({
			items: this.state.items
		});
	}

	handlePress = () => {
		this.setState({
			items: this.state.items.concat([
				{
					text: '',
					checked: false,
				}
			])
		});
	}

	handleCheckItem = (index) => {
		this.state.items[index].checked = !(this.state.items[index].checked);
		this.setState({
			items: this.state.items
		});
	}

	render() {
		return (
			<View style={styles.container}>
				<DateHeader testID='date-header'/>
				<ItemList testID='item-list' items={this.state.items} onChange={this.handleChange} onCheckItem={this.handleCheckItem} />
				<AddItemButton testID='add-item-button' onPress={this.handlePress} />
			</View>
		);
	}
}

const styles = StyleSheet.create({
	container: {
		height: 800,
		padding: 50,
		justifyContent: 'space-between',
	},
});

