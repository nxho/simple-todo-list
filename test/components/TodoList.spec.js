import React from 'react';
import { render, shallow, waitForElement, fireEvent } from 'react-native-testing-library';
import TestRenderer from 'react-test-renderer';
import {
  ScrollView,
  TextInput,
  TouchableOpacity,
  View,
} from 'react-native';
import TodoList from '../../components/TodoList';

describe('TodoList component', () => {
  it('renders correctly', () => {
    const { output } = shallow(<TodoList />);

    expect(output).toMatchSnapshot();
  });

  xit('pressing add item button creates new item that is editable', async () => {
    const { getByType, getByTestId, update, debug } = render(<TodoList />);

    fireEvent.press(getByTestId('add-item-button'));

    // TODO: this currently breaks because react-test-renderer cannot render ScrollView????
    // it works if you change it to a View component instead
    await expect(waitForElement(() => getByTestId('item0-text-input'))).resolves.not.toThrow();

    await fireEvent.changeText(getByTestId('item0-text-input'), 'List item 0');
    expect(getByTestId('item0-text-input').props.value).toEqual('List item 0');
  });

  it('pressing add item button creates new item that is editable - test-renderer version', () => {
    const renderer = TestRenderer.create(<TodoList />);
    const root = renderer.root;

    root.findByType(TouchableOpacity).props.onPress();

    expect(root.findByType(ScrollView).props.children[0]).toBeDefined();
    expect(root.findByType(ScrollView).props.children[0].props.testID).toEqual('item0-text-input');
    expect(root.findByType(ScrollView).props.children[0].props.value).toEqual('');

    root.findByType(ScrollView).props.children[0].props.onChangeText('List item 0');
    expect(root.findByType(ScrollView).props.children[0].props.value).toEqual('List item 0');
  });
});
