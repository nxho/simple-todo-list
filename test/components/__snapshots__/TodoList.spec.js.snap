// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component renders correctly 1`] = `
<Component
  style={
    Object {
      "height": 800,
      "justifyContent": "space-between",
      "padding": 50,
    }
  }
>
  <DateHeader
    testID="date-header"
  />
  <ItemList
    items={Array []}
    onChange={[Function]}
    onCheckItem={[Function]}
    testID="item-list"
  />
  <AddItemButton
    onPress={[Function]}
    testID="add-item-button"
  />
</Component>
`;
